casper/src/main/scala/coop/rchain/casper/TuranOracle.scala

```diff
@@ -0,0 +1,103 @@
+ * tolerate up to fault_tolerance_threshold fraction of the total weight to equivocate.
+ *
+ * In the extreme case when your fault tolerance threshold is 1,
+ * all validators must be part of the clique that supports the estimate in order to state that it is finalized.
+ * If all validators are indeed part of the clique, min_max_clique_weight will hopefully be equal to total_weight
+ * and is_safe will reduce to total_weight >= total_weight and evaluate to true.
+ */
+class TuranOracle(blocks: collection.Map[ByteString, BlockMessage],
+                  estimate: BlockMessage,
+                  latestBlocks: collection.Map[ByteString, BlockMessage],
+                  fault_tolerance_threshold: Float) {
+  def totalWeight(): Int = weightMapTotal(estimate)
+  def isSafe: Boolean = {
+    val fault_tolerance = 2 * min_max_clique_weight() - totalWeight()
+    fault_tolerance >= fault_tolerance_threshold * totalWeight()
+  }
+
+  private def min_max_clique_weight(): Int =
+    // To have a maximum clique of half the total weight, you need at least twice the weight of the candidates to be greater than the total weight
+    if (2 * candidates().values.sum < totalWeight()) {
+      0
+    } else {
+      val vertexCount = candidates().keys.size
+      val edgeCount   = agreementGraphEdgeCount(blocks, estimate, latestBlocks, candidates())
+      minTotalValidatorWeight(estimate, maxCliqueMinSize(vertexCount, edgeCount))
+    }
+
+  private def candidates(): Map[ByteString, Int] = {
+    val estimateMainParent = mainParent(blocks, estimate)
+    val weights = estimateMainParent match {
+      case Some(parent) => weightMap(parent)
+      case None         => weightMap(estimate) // Genesis
+    }
+    for {
+      (validator, stake) <- weights
+      latestBlock        <- latestBlocks.get(validator) if compatible(estimate, latestBlock)
+    } yield (validator, stake)
+  }
+
+  // TODO: Add free messages
+  def agreementGraphEdgeCount(blocks: collection.Map[ByteString, BlockMessage],
+                              estimate: BlockMessage,
+                              latestBlocks: collection.Map[ByteString, BlockMessage],
+                              candidates: Map[ByteString, Int]): Int = {
+    def seesAgreement(that: ByteString, other: ByteString): Boolean =
+      (for {
+        selfLatest <- latestBlocks.get(that).toList
+        justification <- selfLatest.justifications
+                          .map {
+                            case Justification(creator: ByteString, latestBlock: ByteString) =>
+                              creator -> latestBlock
+                          }
+                          .toMap
+                          .values
+                          .toList
+        justificationBlock <- blocks.get(justification)
+        if justificationBlock.sig == other && compatible(estimate, justificationBlock)
+      } yield justificationBlock).size == 1
```

Comment:
I'm not totally following all the conversions that are going on here, but you seem to be turning a lot of `Option`s into `List`s and then checking at the end if the length of the list is 1. Would it make sense to leave the `Option`s alone and check `isDefined` at the end? Or phrased another way, are you ever expecting the final `List` to have more than one element?

Reply:
Resolved through chat